### PR TITLE
Server gun prefix requirement feature

### LINKED ISSUE #608: create a server config to require a prefix for all keys
If this config is specified, any request to get or delete a GUN without the prefix should provide a specific error that says something like "this name is invalid because it lacks the prefix X".

### PR DESCRIPTION
**Pull request by cyli**
Fixes #608

### CODE REVIEW

**Comment by endophage** [comment]
This implementation works for me. If we want the behavior to be more complicated in the future we can discuss it when we have a better idea what it should be.

**Comment by endophage** [comment]
Let's add the additional config entry to direct how to use the prefixes to this PR, but only implement `"restrict"` (i.e. the current proposed behaviour).

Another PR can then implement `"trim"` (or whatever we decide to call the other option).

**Comment by cyli** [comment]
@endophage:  Hmm... do you want the user to actually have to specify the `"restrict"` behavior?  Since that is the default behavior, not sure if we need another configuration option yet.

When we implement `"trim"`, can that be a boolean option that gets added to the `"repositories"` section?  Do we ever want to trim without restricting?

**Comment by cyli** [comment]
Update: after IRL discussion, we determined that variable behavior can be added later with a `behavior` config option or `trim` flag or some such.  Right now we're leaving it off because it is the default behavior, and we're not sure if we want others yet.

File: server/errors/errors.go [hunk 1]
```diff
@@ -83,8 +83,14 @@ var (
 	ErrNoKeyAlgorithm = errcode.Register(errGroup, errcode.ErrorDescriptor{
 		Value:          "NO_KEYALGORITHM",
 		Message:        "The server does not have a key algorithm configured.",
-		Description:    "No key algorihtm has been configured for the server and it has been asked to perform an operation that requires generation.",
+		Description:    "No key algorithm has been configured for the server and it has been asked to perform an operation that requires generation.",
```
Comment (riyazdf) [inline server/errors/errors.go]: thanks for the fix 👍

File: cmd/notary-server/config.go [hunk 2]
```diff
@@ -21,9 +23,21 @@ import (
...
+	gorethink "gopkg.in/dancannon/gorethink.v2"
 )
 
+// gets the required gun prefixes accepted by this server
+func getRequiredGunPrefixes(configuration *viper.Viper) ([]string, error) {
+	prefixes := configuration.GetStringSlice("repositories.gun_prefixes")
+	for _, prefix := range prefixes {
+		p := path.Clean(strings.TrimSpace(prefix))
+		if p+"/" != prefix || strings.HasPrefix(p, "/") || strings.HasPrefix(p, "..") {
+			return nil, fmt.Errorf("Invalid GUN prefix %s", prefix)
```
Comment (riyazdf) [inline cmd/notary-server/config.go]: nit: do we want start the error message with lowercase?

File: cmd/notary-server/config.go [hunk 3]
```diff
@@ -21,9 +23,21 @@ import (
...
-	"gopkg.in/dancannon/gorethink.v2"
+	gorethink "gopkg.in/dancannon/gorethink.v2"
 )
 
+// gets the required gun prefixes accepted by this server
+func getRequiredGunPrefixes(configuration *viper.Viper) ([]string, error) {
+	prefixes := configuration.GetStringSlice("repositories.gun_prefixes")
+	for _, prefix := range prefixes {
+		p := path.Clean(strings.TrimSpace(prefix))
+		if p+"/" != prefix || strings.HasPrefix(p, "/") || strings.HasPrefix(p, "..") {
```
Comment (riyazdf) [inline cmd/notary-server/config.go]: I'm wondering if we also want to enforce the leading `/` and `..` cases for GUNs if the config does not specify any prefixes?

Unless I'm missing something, notary could still create trust data for GUNs like `/gun` and `../../gun` if we never get into this loop

Comment (riyazdf) [inline cmd/notary-server/config.go]: There's some pretty varied behavior for this currently in master, not sure if we want to separate fixing this into another PR.  Tested with notary from master and compose setup:

leading `/`:
```
$ bin/notary -c cmd/notary/config.json init /test/
Root key found, using: 2930402da13b3062f9f59cb3bc6f93d04ae0caaf5778318cf2de8a2eb123367b
Enter passphrase for root key with ID 2930402:
Enter passphrase for new targets key with ID 8ec423d (/test):
Repeat passphrase for new targets key with ID 8ec423d (/test):

* fatal: unable to reach trust server at this time: 301.
```

one level `..` escape:
```
$ bin/notary -c cmd/notary/config.json init ../test
Root key found, using: 2930402da13b3062f9f59cb3bc6f93d04ae0caaf5778318cf2de8a2eb123367b
Enter passphrase for root key with ID 2930402:
Enter passphrase for new targets key with ID 8fa72b4 (../test):
Repeat passphrase for new targets key with ID 8fa72b4 (../test):

* fatal: timestamp key trust data unavailable.  Has a notary repository been initialized?
```

multiple `..` escape
```
$ bin/notary -c cmd/notary/config.json init ../../test
Root key found, using: 2930402da13b3062f9f59cb3bc6f93d04ae0caaf5778318cf2de8a2eb123367b
Enter passphrase for root key with ID 2930402:
Enter passphrase for new targets key with ID c36d72b (../../test):
Repeat passphrase for new targets key with ID c36d72b (../../test):

* fatal: failed to add key to filestore: path outside file store
```

Comment (cyli) [inline cmd/notary-server/config.go]: I was wondering that too - https://github.com/docker/notary/issues/742 tracks the issue, but we can totally check it here.  I think @endophage is testing out whether the Go http server automatically collapses the URL for you too (e.g. whether the Go server collapses `http://localhost:8080/foo/../../../bar` to `http://localhost:8080/bar`)

Comment (cyli) [inline cmd/notary-server/config.go]: We should probably also validate the GUN when `init`-ing using the CLI, because the user can probably mess up their local file system.

Comment (cyli) [inline cmd/notary-server/config.go]: Oh duh sorry they can't pollute on init :).  I guess not super important then, but may just be nice to have a nicer error message about invalid GUN though

File: cmd/notary-server/config.go (hunk 2 above)
Comment (cyli) [inline cmd/notary-server/config.go]: Yes!  Thanks, forgot that

File: cmd/notary-server/config.go (hunk 3 above)
Comment (riyazdf) [inline cmd/notary-server/config.go]: totally agreed, a separate PR makes sense

**Comment by riyazdf** [comment]
LGTM after small caps nit 👍

File: cmd/notary-server/config.go (hunk 3 above)
Comment (endophage) [inline cmd/notary-server/config.go]: So those filepath escapes get cleaned out and a 301 returned at the Go http server level, we never see them. Would be good to add checks to the CLI in a different PR. Because our server code won't even see them, doesn't make sense to put that in this PR.

**Comment by endophage** [comment]
LGTM
